go-controller/pkg/ovn/egressgw.go

```diff
@@ -83,15 +86,10 @@ func (oc *Controller) ensureRouteInfoLocked(podName ktypes.NamespacedName) (*ext
 
 	// 5. If routeInfo originally existed in the cache, there is a chance it was altered after we
 	// released the cache lock, and before we were able to lock routeInfo
-	if routeInfoExisted {
-		// Check that the routeInfo wasn't altered in the cache while we were waiting for the lock
-		// and that routeInfo that we locked is not stale.
-		oc.exGWCacheMutex.RLock()
-		defer oc.exGWCacheMutex.RUnlock()
-		if routeInfo != oc.externalGWCache[podName] {
-			routeInfo.Unlock()
-			return nil, fmt.Errorf("routeInfo for pod %s, was altered during ensure route info", podName)
-		}
+	if routeInfo.deleted {
```

Comment:
since we only care about whether it got deleted or not, we should be ok in not wanting to check in any other edits I guess.

Reply:
Yeah, and this part is the tradeoff with no longer deadlocking. If we get to this point, we'll return the error and that means `addGWRoutesForPod()` will fail, and won't add the pod on a quick create/delete/create cycle. The risk is that when the object is deleted we grab the routeInfo lock, set deleted=true, start the goroutine, and return the locked routeInfo, but before the goroutine is actually able to begin executing the pod gets re-added and `ensureRouteInfoLocked()` gets called. It grabs the to-be-deleted routeinfo out of the cache (since the `deleteRouteInfoLocked` goroutine hasn't started yet), locks it, and sees that it's deleted, and returns the error, failing to reach consistent state if no further pod events occur.

But that's better than a deadlock since we can work around it by deleting, waiting a couple seconds, and recreating the pod, IMO.